README.md
<p align="center">
  <a href="https://github.com/Fliggy-Android-Team">
    <img width="200" src="https://gw.alicdn.com/tfs/TB1a288sxD1gK0jSZFKXXcJrVXa-360-360.png">
  </a>
</p>

<h1 align="center">FSwitch</h1>


<div align="center">

<p>An interesting and practical switch component.</p>

<p>Supports setting tips, slider decorations, shadows, and good interaction.</p>

<p><strong>Author：<a href="https://github.com/chenBingX">Newton</a>(<a href="coorchice.cb@alibaba-inc.com">coorchice.cb@alibaba-inc.com</a>)</strong></p>

<p>

<a href="https://pub.dev/packages/fswitch#-readme-tab-">
    <img height="20" src="https://img.shields.io/badge/Version-1.1.2-important.svg">
</a>


<a href="https://github.com/Fliggy-Android-Team/fswitch">
    <img height="20" src="https://img.shields.io/badge/Build-passing-brightgreen.svg">
</a>


<a href="https://github.com/Fliggy-Android-Team">
    <img height="20" src="https://img.shields.io/badge/Team-FAT-ffc900.svg">
</a>

<a href="https://www.dartcn.com/">
    <img height="20" src="https://img.shields.io/badge/Language-Dart-blue.svg">
</a>

<a href="https://pub.dev/documentation/fswitch/latest/fswitch/fswitch-library.html">
    <img height="20" src="https://img.shields.io/badge/API-done-yellowgreen.svg">
</a>

<a href="http://www.apache.org/licenses/LICENSE-2.0.txt">
   <img height="20" src="https://img.shields.io/badge/License-Apache--2.0-blueviolet.svg">
</a>

<p>
<p>

<img height="500" src="https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fswitch/fswitch_all.png">

</div>

**English | [简体中文](https://github.com/Fliggy-Android-Team/fswitch/blob/master/README_CN.md)**

> Like it? Please cast your **Star**  🥰 ！

# ✨ Features

- Support opening and closing tips

- Support setting decoration for **Slider**

- Support flexible configuration available state

- More sense of space **Shadow**

- Good switch interactive experience

# 🛠 Guide

## ⚙️ Parameters

|Param|Type|Necessary|Default|desc|
|---|---|:---:|---|---|
|open|bool|false|false|Whether it is open. The default is false.|
|onChanged|ValueChanged<bool>|true|null|This function will be called back when the switch state changes.|
|width|double|false|59.23|width. Default 59.23, in line with aesthetics 😃|
|height|double|false|null|height. By default, it will be calculated according to [width], which is in line with aesthetics 😃|
|offset|double|false|null|Distance between slider and edge|
|openChild|Widget|false|null|Prompt style of open state|
|closeChild|Widget|false|null|Prompt style of closed state|
|childOffset|double|false|null|Tip to edge distance|
|color|Color|false|null|Background color when off|
|openColor|Color|false|null|Background color when open|
|sliderColor|Color|false|null|Slider color|
|enable|bool|false|true|it's usable or not|
|sliderChild|Widget|false|null|The component in the slider. Beyond the range will be cropped.|
|shadowColor|Color|false|null|Set component shadow color|
|shadowOffset|Offset|false|null|Set component shadow offset|
|shadowBlur|double|false|0.0|Set the standard deviation of the component Gaussian and shadow shape convolution|

## 📺 Demo

### 🔩 Basic Demo

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fswitch/fswitch_normal.gif)

```dart
FSwitch(
  onChanged: (v) {},
)
```

**FSwitch** needs to always set `onChanged` in order to respond to the switch status.

### Available status & shadow

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fswitch/fswitch_enable_shadow.gif)

```dart
FSwitch(
  onChanged: (bool value) {
    value_1 = value;
  },
  open: value_1,
  enable: enable_1,
  shadowColor: Colors.black.withOpacity(0.5),
  shadowBlur: 3.0,
),
```

You can change the available status of **FSwitch** through `enable`. Once **FSwitch** becomes available, the switch state cannot be changed. 

In addition, **FSwitch** provides flexible and convenient shadow configuration support.

### 🌝 Prompt

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fswitch/fswitch_tip.gif)

```dart

/// #1
FSwitch(
  width: 65.0,
  height: 35.538,
  onChanged: (v) {},
  closeChild: Text(
    "Off",
    style: TextStyle(
        color: Colors.white, fontSize: 11),
  ),
  openChild: Text(
    "On",
    style: TextStyle(
        color: Colors.white, fontSize: 11),
  ),
),

/// #2
FSwitch(
  open: true,
  onChanged: (v) {},
  closeChild: Icon(
    Icons.close,
    size: 16,
    color: Colors.white,
  ),
  openChild: Icon(
    Icons.check,
    size: 16,
    color: Colors.white,
  ),
),

/// #3
FSwitch(
  width: 65.0,
  height: 35.538,
  onChanged: (v) {},
  closeChild: Text(
    "😒",
    style: TextStyle(fontSize: 20),
  ),
  openChild: Text(
    "😃",
    style: TextStyle(fontSize: 20),
  ),
  childOffset: 3.0,
),
```

The open and closed status indicators can be set for **FSwitch** through `openChild` and` closeChild`, respectively.


### 👀 Slider decoration

![](https://raw.githubusercontent.com/chenBingX/img/master/Flutter/fswitch/fswitch_slider.gif)

```dart
FSwitch(
  width: 300,
  height: 38,
  onChanged: (bool value) {},
  sliderChild: Text(
    "😃",
    style: TextStyle(fontSize: 20),
  ),
)
```

`sliderChild` will allow developers to set` child` decorations for **Slider**.

It's interesting!


# 😃 How to use？

Add dependencies in the project `pubspec.yaml` file:

## 🌐 pub dependency

```
dependencies:
  fswitch: ^<version number>
```

> ⚠️ Attention，please go to [**pub**](https://pub.dev/packages/fswitch) to get the latest version number of **FSwitch**

## 🖥 git dependencies

```
dependencies:
  fswitch:
    git:
      url: 'git@github.com:Fliggy-Android-Team/fswitch.git'
      ref: '<Branch number or tag number>'
```


> ⚠️ Attention，please refer to [**FSwitch**](https://github.com/Fliggy-Android-Team/fswitch) official project for branch number or tag.


# 💡 License

```
Copyright 2020-present Fliggy Android Team <alitrip_android@list.alibaba-inc.com>.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at following link.

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

```


### Like it? Please cast your [**Star**](https://github.com/Fliggy-Android-Team/fswitch) 🥰 ！



CHANGELOG.md
## 1.1.0

- Support shadow

- Support configuration **Slider** decoration

- Improve interactivity

- Optimization function

## 1.1.1 

- Upgrade configuration


## 1.1.2

- Upgrade configuration


test/fswitch_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:fswitch/fswitch.dart';

void main() {

}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:fswitch_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


readme.py
# coding=utf-8
import os
import sys


def print_error(s):
    print ('\033[0;31mError: {content}\033[0m'.format(content=s))


def print_color(s, mode='0', color='37'):
    return '\033[{mode};{color}m{content}\033[0m'.format(mode=mode, color=color, content=s)


# text是否包含content
def is_contain(text, content):
    text = str(text)
    content = str(content)
    if text.find(content) != -1:
        return True
    else:
        return False


def is_empty(s):
    return s is None or len(s) == 0


def runCmd(cmd):
    # print (cmd)
    f = os.popen(cmd)
    text = f.read()
    f.close()
    return text



def get_name_in_yaml(yaml_path):
    name = ''
    with open(yaml_path, mode='r') as projects_yaml:
        lines = projects_yaml.readlines()
        for line in lines:
            if line.startswith("name:"):
                name = line.strip().lstrip().rstrip(',').replace(' ', '').split(':')[1]
                break
    return name


def create_readme():
    readme_path = 'README.md'
    yaml_path = 'pubspec.yaml'
    if not os.path.exists(yaml_path):
        print_error("项目缺少 pubspec.yaml 配置文件，请检查项目配置文件是否存在或路径是否正确")
        return
    with open(readme_path, 'w+') as readme:
        project_name = get_name_in_yaml(yaml_path)
        lines = []
        lines.append("# " + project_name)
        lines.append("\n\n")
        lines.append("# fapi/" + project_name)
        lines.append("\n\n")
        lines.append("![](https://badgen.net/badge/tnpm/请自行填写-组件最新tag版本号/blue)")
        lines.append("\n\n")
        lines.append("> [使用前请参考【Flutter组件开发使用文档】](https://yuque.antfin-inc.com/hummingbird/flutter-guide/sv9shr)")
        lines.append("\n\n")
        lines.append("* 主理人：请自行填写-名称 <请自行填写-邮箱>")
        lines.append("\n\n")
        lines.append("# 设计规范")
        lines.append("\n\n")
        lines.append('<img src="请自行填写-设计稿图片" />')
        lines.append("\n\n")
        lines.append("# 使用指南")
        lines.append("\n\n")
        lines.append("## 参数")
        lines.append("\n\n")
        lines.append("|参数|类型|必要|默认值|说明|")
        lines.append("\n")
        lines.append("|---|---|:---:|---|---|")
        lines.append("\n")
        lines.append("|width|double|否|null|宽度|")
        lines.append("\n")
        lines.append("|**|**|**|**|**|")
        lines.append("\n")
        lines.append("|**|**|**|**|**|")
        lines.append("\n")
        lines.append("|**|**|**|**|**|")
        lines.append("\n\n")
        lines.append("## 使用示例")
        lines.append("\n\n")
        lines.append("```")
        lines.append("\n\n")
        lines.append("***")
        lines.append("\n")
        lines.append("***")
        lines.append("\n")
        lines.append("***")
        lines.append("\n\n")
        lines.append("```")
        lines.append("\n\n")
        lines.append("[【传送门】Dartpad - {widget} 演示](https://dartpad.dev/{gist})".format(widget=project_name, gist='请自行填写-gist代码地址'))
        lines.append("\n\n")
        lines.append('<iframe name="dartpad" width=90% height=1000 frameborder=0 src="https://dartpad.dev/请自行填写-gist代码地址"></iframe>')
        lines.append("\n\n")
        lines.append("# 历史版本")
        lines.append("\n\n")
        lines.append("[【传送门】组件历史版本](http://gitlab.alibaba-inc.com/fapi/{widget}/tags)".format(widget=project_name))
        lines.append("\n\n")

        readme.seek(0)
        readme.truncate()
        readme.writelines(lines)
        print(print_color(s="README.md 模版生成完成！", color='32'))


create_readme()


lib/fswitch.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

/// 一个有趣的、实用的开关组件。支持设置提示、滑块装饰、阴影以及良好的交互。
///
/// An interesting and practical switch component.
/// Supports setting tips, slider decorations, shadows, and good interaction.
// ignore: must_be_immutable
class FSwitch extends StatefulWidget {
  /// 是否处于打开状态。默认 false。
  ///
  /// Whether it is open. The default value is false.
  bool open;

  /// 当开关状态发生改变时，会回调该函数。
  ///
  /// This function will be called back when the switch state changes.
  ValueChanged<bool> onChanged;

  /// 宽度。默认 59.23，符合美学 😃
  ///
  /// width. Default 59.23, in line with aesthetics 😃
  double width;

  /// 高度。默认会根据 [width] 进行计算，符合美学 😃
  ///
  /// height. By default, it will be calculated according to [width], which is in line with aesthetics 😃
  double height;

  /// 滑块与边缘的间距
  ///
  /// Distance between slider and edge
  double offset;

  /// 打开状态的提示样式
  ///
  /// Prompt style of open state
  Widget openChild;

  /// 关闭状态的提示样式
  ///
  /// Prompt style of closed state
  Widget closeChild;

  /// 提示与边缘的间距
  ///
  /// Prompt to edge distance
  double childOffset;

  /// 关闭状态下的背景色
  ///
  /// Background color when off
  Color color;

  /// 打开状态下的背景色
  ///
  /// Background color when open
  Color openColor;

  /// 滑块颜色
  ///
  /// Slider color
  Color sliderColor;

  /// 滑块中的组件。超过范围会被剪裁。
  ///
  /// Components in the slider。Beyond the range will be cropped。
  Widget sliderChild;

  /// 是否可用
  ///
  /// it's usable or not
  bool enable;

  /// 设置组件阴影颜色
  ///
  /// Set component shadow color
  Color shadowColor;

  /// 设置组件阴影偏移
  ///
  /// Set component shadow offset
  Offset shadowOffset;

  /// 设置组件高斯与阴影形状卷积的标准偏差。
  ///
  /// Sets the standard deviation of the component's Gaussian convolution with the shadow shape.
  double shadowBlur;

  FSwitch({
    Key key,
    @required this.onChanged,
    this.open = false,
    this.width = 59.23,
    this.height,
    this.offset,
    this.childOffset,
    this.closeChild,
    this.openChild,
    this.color,
    this.openColor,
    this.sliderColor,
    this.sliderChild,
    this.enable = true,
    this.shadowColor,
    this.shadowOffset,
    this.shadowBlur = 0.0,
  })  : assert(open != null && onChanged != null,
            "open and onChanged can't be None!"),
        super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _FSwitch();
  }
}

class _FSwitch extends State<FSwitch> {
  double fixOffset;
  bool draging = false;
  double dragDxW = 10.0;

  @override
  void initState() {
    super.initState();
    fixOffset = widget.open
        ? widget.width -
            (widget.offset ??
                    2.0 / 36.0 * ((widget.height ?? widget.width * 0.608))) *
                2.0 -
            (widget.height ?? widget.width * 0.608) * (32.52 / 36.0)
        : 0;
  }

  @override
  void didUpdateWidget(FSwitch oldWidget) {
    super.didUpdateWidget(oldWidget);
    fixOffset = widget.open
        ? widget.width -
            (widget.offset ??
                    2.0 / 36.0 * ((widget.height ?? widget.width * 0.608))) *
                2.0 -
            (widget.height ?? widget.width * 0.608) * (32.52 / 36.0)
        : 0;
  }

  @override
  Widget build(BuildContext context) {
    double height = widget.height ?? widget.width * 0.608;
    double circleSize = (height * (32.52 / 36.0));
    widget.offset = widget.offset ?? 2.0 / 36.0 * height;
    double childOffset = widget.childOffset ?? height / 5.0;
    widget.color = widget.color ?? Color(0xffcccccc);
    widget.openColor = widget.openColor ?? Color(0xffffc900);

    List<Widget> children = [];

    /// background
    var showShadow = widget.shadowColor != null && widget.shadowBlur != 0;
    var background = AnimatedContainer(
      duration: Duration(milliseconds: 350),
      decoration: BoxDecoration(
        color: (widget.open ? widget.openColor : widget.color) ?? widget.color,
        borderRadius: BorderRadius.all(Radius.circular(height / 2.0)),
        boxShadow: showShadow
            ? [
                BoxShadow(
                  color: widget.shadowColor,
                  offset: widget.shadowOffset ?? Offset(0, 0),
                  blurRadius: widget.shadowBlur,
                )
              ]
            : null,
      ),
      child: Container(
        width: widget.width,
        height: height,
      ),
    );
    children.add(background);

    /// Prompt
    var showChild = widget.open ? widget.openChild : widget.closeChild;
    if (showChild != null) {
      showChild = Positioned(
        left: widget.open ? childOffset : null,
        right: widget.open ? null : childOffset,
        child: showChild,
      );
      children.add(showChild);
    }

    /// slider
    var slider = AnimatedContainer(
      margin: EdgeInsets.fromLTRB(widget.offset + fixOffset, 0, 0, 0),
      duration: Duration(milliseconds: 200),
      width: circleSize + (draging ? dragDxW : 0.0),
      child: Container(
        height: circleSize,
        clipBehavior: Clip.antiAlias,
        alignment: Alignment.center,
        decoration: BoxDecoration(
            color: widget.sliderColor ?? Color(0xffffffff),
            borderRadius: BorderRadius.all(Radius.circular(circleSize / 2.0))),
        child: widget.sliderChild,
      ),
    );
    children.add(slider);

    /// When in an unavailable state, add a mask
    if (!widget.enable) {
      var disableMask = Opacity(
        opacity: 0.6,
        child: Container(
          width: widget.width,
          height: height,
          decoration: BoxDecoration(
              color: Color(0xfff1f1f1),
              borderRadius: BorderRadius.all(Radius.circular(height / 2.0))),
        ),
      );
      children.add(disableMask);
    }

    return GestureDetector(
      onTap: widget.enable ? _handleOnTap : null,
      onHorizontalDragEnd: widget.enable ? _handleOnHorizontalDragEnd : null,
      onHorizontalDragUpdate:
          widget.enable ? _handleOnHorizontalDragUpdate : null,
      onHorizontalDragCancel: widget.enable ? _handleDragCancel : null,
      onHorizontalDragStart: widget.enable ? _handleDragStart : null,
      child: Container(
        child: Stack(
          alignment: Alignment.centerLeft,
          children: children,
        ),
      ),
    );
  }

  void _handleOnTap() {
    setState(() {
      widget.open = !widget.open;
      double height = widget.height ?? widget.width * 0.608;
      double circleSize = (height * (32.52 / 36.0));
      if (widget.open) {
        fixOffset = widget.width - widget.offset - circleSize - widget.offset;
      } else {
        fixOffset = 0;
      }
      widget.onChanged(widget.open);
    });
  }

  void _handleDragStart(DragStartDetails details) {
    setState(() {
      draging = true;
    });
  }

  void _handleOnHorizontalDragUpdate(DragUpdateDetails details) {
    setState(() {
      double height = widget.height ?? widget.width * 0.608;
      double circleSize = (height * (32.52 / 36.0));
      fixOffset = fixOffset + details.delta.dx;
      if (fixOffset < 0) {
        fixOffset = 0;
      } else if (fixOffset >
          widget.width -
              widget.offset -
              circleSize -
              (draging ? dragDxW : 0.0) -
              widget.offset) {
        fixOffset = widget.width -
            widget.offset -
            circleSize -
            (draging ? dragDxW : 0.0) -
            widget.offset;
      }
    });
  }

  void _handleOnHorizontalDragEnd(DragEndDetails details) {
    setState(() {
      draging = false;
      double height = widget.height ?? widget.width * 0.608;
      double circleSize = (height * (32.52 / 36.0));
      double center = (widget.width -
              widget.offset -
              circleSize -
              (draging ? dragDxW : 0.0) -
              widget.offset) /
          2;
      bool cacheValue = widget.open;
      if (fixOffset < center) {
        fixOffset = 0;
        widget.open = false;
      } else {
        fixOffset = center * 2;
        widget.open = true;
      }
      if (cacheValue != widget.open) {
        widget.onChanged(widget.open);
      }
    });
  }

  void _handleDragCancel() {
    setState(() {
      draging = false;
    });
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:fswitch/fswitch.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: FSwitchPage(),
    );
  }
}

class FSwitchPage extends StatefulWidget {
  static const String routeName = '/fswitch';

  @override
  State<FSwitchPage> createState() => _FSwitchPageState();
}

class _FSwitchPageState extends State<FSwitchPage> {
  var enable_1 = true;
  var value_1 = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Color(0xffffc900),
        title: const Text('FSwitch'),
        centerTitle: true,
      ),
      body: SingleChildScrollView(
        child: Container(
          padding: EdgeInsets.fromLTRB(0, 0, 0, 0),
          child: Column(
            children: <Widget>[
              /// FSwitch
              buildTitle("FSwitch"),
              buildSmallMargin(),
              buildFSwitch(),
              buildSmallMargin(),
              buildDesc("一个简单的 FSwitch"),

              /// FSwitch with child
              buildTitle("FSwitch with child"),
              buildSmallMargin(),
              buildFSwitchWithChild(),
              buildSmallMargin(),
              buildDesc("FSwitch 支持自定义提示组件\n"
                  "通过 unselectedChild 和 selectedChild 属性"),
              _buildMiddleMargin(),

              /// Disable FSwitch
              buildTitle("Disable FSwitch"),
              buildSmallMargin(),
              buildDisableFSwitch(),
              buildDesc("通过 enable 属性可以设置 FSwitch 的可用状态"),
              _buildMiddleMargin(),

              /// Try Drag？
              buildTitle("Try Drag？"),
              buildSmallMargin(),
              buildDragFSwitch(),
              buildSmallMargin(),
              buildDesc("FSwitch 除了支持点击切换状态，也支持拖拽手势"),
            ],
          ),
        ),
      ),
    );
  }

  FSwitch buildDragFSwitch() {
    return FSwitch(
      width: 300,
      height: 38,
      onChanged: (bool value) {},
      sliderChild: Text(
        "😃",
        style: TextStyle(fontSize: 20),
      ),
    );
  }

  StatefulBuilder buildDisableFSwitch() {
    return StatefulBuilder(
      builder: (context, setState) {
        return Column(
          children: <Widget>[
            Text(
              enable_1 ? "可用状态" : "不可用状态",
              style: TextStyle(color: Colors.grey, fontSize: 12),
            ),
            buildSmallMargin(),
            FSwitch(
              onChanged: (bool value) {
                value_1 = value;
              },
              open: value_1,
              enable: enable_1,
              shadowColor: Colors.black.withOpacity(0.5),
              shadowBlur: 3.0,
            ),
            buildSmallMargin(),
            RaisedButton(
              onPressed: () {
                setState(() {
                  enable_1 = !enable_1;
                });
              },
              child: Text("Toggle disable"),
              color: Color(0xffffc900),
            )
          ],
        );
      },
    );
  }

  Container buildFSwitchWithChild() {
    return Container(
      width: 300,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: <Widget>[
          FSwitch(
            width: 65.0,
            height: 35.538,
            onChanged: (v) {},
            closeChild: Text(
              "Off",
              style: TextStyle(color: Colors.white, fontSize: 11),
            ),
            openChild: Text(
              "On",
              style: TextStyle(color: Colors.white, fontSize: 11),
            ),
          ),
          FSwitch(
            open: true,
            onChanged: (v) {},
            closeChild: Icon(
              Icons.close,
              size: 16,
              color: Colors.white,
            ),
            openChild: Icon(
              Icons.check,
              size: 16,
              color: Colors.white,
            ),
          ),
          FSwitch(
            width: 65.0,
            height: 35.538,
            onChanged: (v) {},
            closeChild: Text(
              "😒",
              style: TextStyle(fontSize: 20),
            ),
            openChild: Text(
              "😃",
              style: TextStyle(fontSize: 20),
            ),
            childOffset: 3.0,
          ),
        ],
      ),
    );
  }

  FSwitch buildFSwitch() {
    return FSwitch(
      onChanged: (v) {},
    );
  }

  SizedBox _buildMiddleMargin() {
    return const SizedBox(
      height: 18,
    );
  }

  SizedBox buildSmallMargin() {
    return const SizedBox(
      height: 10,
    );
  }

  Padding buildDesc(String desc) {
    return Padding(
        padding: const EdgeInsets.all(8),
        child: Text(
          desc,
          textAlign: TextAlign.center,
          style: TextStyle(
            color: Colors.grey,
            fontSize: 12,
          ),
        ));
  }

  Container buildTitle(String title) {
    return Container(
      alignment: Alignment.centerLeft,
      padding: EdgeInsets.all(9),
      color: Color(0xffe0e0e0).withOpacity(0.38),
      child: Text(
        title,
        style: TextStyle(color: Colors.grey),
      ),
    );
  }
}
